pip/vcs/git.py

```diff
@@ -74,27 +74,34 @@ def export(self, location):
                 show_stdout=False, cwd=temp_dir.path
             )
 
-    def check_rev_options(self, rev, dest, rev_options):
-        """Check the revision options before checkout to compensate that tags
-        and branches may need origin/ as a prefix.
-        Returns the SHA1 of the branch or tag if found.
+    def get_ref_sha(self, dest, ref):
         """
-        revisions = self.get_short_refs(dest)
-
-        origin_rev = 'origin/%s' % rev
-        if origin_rev in revisions:
-            # remote branch
-            return [revisions[origin_rev]]
-        elif rev in revisions:
-            # a local tag or branch name
-            return [revisions[rev]]
-        else:
+        Return a commit hash for the reference if the given name is a ref.
+        Otherwise, return None.
+
+        Args:
+          dest: the repository directory.
+          ref: the ref name.
+        """
+        output = self.run_command(['show-ref', '--hash', ref], cwd=dest,
+                                  show_stdout=False, on_returncode='ignore')
+        # The git-show-ref command can return multiple references, so we
+        # return only the first.
+        for sha in output.strip().splitlines():
+            return sha
```

Comment:
Entirely possible that I'm missing something, but wouldn't this make it ambiguous which remote you're getting if there isn't a local reference? It seems like this assumes that a matching local reference would always appear first in the command output, e.g.:

```
$ git show-ref foo
60bf10ceb7a04c5c73c1e205aedeb0509b09a806 refs/heads/foo
1262759cacb016f590a30d90cfdc7288c1e4d439 refs/remotes/heroku/foo
60bf10ceb7a04c5c73c1e205aedeb0509b09a806 refs/remotes/origin/foo
```

which seems to be correct, however if the local branch doesn't exist:

```
$ git branch -D foo
Deleted branch foo (was 60bf10c).
$ git show-ref foo
1262759cacb016f590a30d90cfdc7288c1e4d438 refs/remotes/heroku/foo
60bf10ceb7a04c5c73c1e205aedeb0509b09a805 refs/remotes/origin/foo
```

The first ref in the list is just due to alphabetical ordering. It seems like in the original function takes care to prefer the `origin` remote, but this change doesn't maintain that.